### PR TITLE
PP-1189: Preemption does not happen if resource in contention is a se…

### PR DESCRIPTION
**Pull request by arungrover**
…rver/queue level resource

#### Issue-ID

* **[PP-1189](https://pbspro.atlassian.net/browse/PP-1189)**

#### Problem description
* Scheduler does not see preempt a low priority job if the resource in contention is a server level resource

#### Cause / Analysis
* There were two problems in this issue - 
 - To generate the list of server/queue level resources scheduler checks if it is a host level resources as well. This means purely server/queue level resources (one with flag - q) does not make it to the list.
- The logic to add server/queue level resource to "resources released list" was flawed.

#### Solution description
* Renamed the resdef_to_check_rassn to resdef_to_check_rassn_select (which is what it is) and created a resdef_to_check_rassn just for server/queue resources. Also changed the logic of creating resource request that is to be released.

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

**Comment by arungrover** [comment]
[test_logs.txt](https://github.com/PBSPro/pbspro/files/1604734/test_logs.txt)

File: src/scheduler/job_info.c [hunk 1]
```diff
@@ -5095,29 +5095,28 @@ nspec **create_res_released_array(status *policy, resource_resv *resresv)
...
  */
-resource_req *create_resreq_rel_list(status *policy, nspec **res_released)
+resource_req *create_resreq_rel_list(status *policy, resource_resv *pjob)
 {
-	int i;
 	resource_req *resreq_rel = NULL;
 	resource_req *rel;
 	resource_req *req;
-	if(policy == NULL || res_released == NULL)
+	if(policy == NULL)
```
Comment (agrawalravi90) [inline src/scheduler/job_info.c]: you probably wanna check that pjob is not NULL before dereferencing it later.

File: src/scheduler/job_info.c [hunk 2]
```diff
@@ -5095,29 +5095,28 @@ nspec **create_res_released_array(status *policy, resource_resv *resresv)
...
-					if(resreq_rel == NULL)
-						resreq_rel = rel;
-				}
+	for(req = pjob->resreq; req != NULL; req = req->next) {
+		if(resdef_exists_in_array(policy->resdef_to_check_rassn, req->def)) {
+			if ((policy->rel_on_susp != NULL) && resdef_exists_in_array(policy->rel_on_susp, req->def) == 0)
+				continue;
+			rel = find_alloc_resource_req(resreq_rel, req->def);
+			if(rel != NULL) {
+				rel->amount += req->amount;
```
Comment (agrawalravi90) [inline src/scheduler/job_info.c]: Don't you need to check if the resource is consumable?

File: src/scheduler/data_types.h [hunk 3]
```diff
@@ -243,6 +243,7 @@ struct status
 	resdef **resdef_to_check;		/* resources to match as definitions */
 	resdef **resdef_to_check_no_hostvnode;	/* resdef_to_check without host/vnode*/
 	resdef **resdef_to_check_rassn;		/* resdef_to_check intersects res_rassn */
+	resdef **resdef_to_check_rassn_select;	/* resdef_to_check intersects res_rassn and host level resource */
```
[the remark this answers is not in the document]
Comment (arungrover) [reply src/scheduler/data_types.h]: Thanks for reviewing the pull @agrawalravi90 
Well in this case comment does mention host level resources as well

File: src/scheduler/job_info.c (hunk 1 above)
Comment (arungrover) [reply src/scheduler/job_info.c]: I'll add that check

File: src/scheduler/job_info.c (hunk 2 above)
Comment (arungrover) [reply src/scheduler/job_info.c]: I think rassn resources are queue/server level resources and they are only consumable. I don't think a resource of type string or array string can be made a rassn resource. So, we are probably ok here.

**Comment by arungrover** [comment]
@agrawalravi90 addressed your comment

File: src/scheduler/job_info.c (hunk 2 above)
Comment (agrawalravi90) [reply src/scheduler/job_info.c]: ah, yes, you are right, thanks for clarifying.

**Comment by agrawalravi90** [comment]
Thanks Arun, the code looks good to me. But Travis failed, you might wanna investigate that.

**Comment by arungrover** [comment]
@agrawalravi90 that build is failing because of some travis or ubuntu issue

**Comment by agrawalravi90** [comment]
Thanks for restarting it, please rebase your branch tomorrow so we can merge it in :)

**Comment by arungrover** [comment]
Thanks @agrawalravi90 and @mike0042  for reviewing the code, I've  rebased the changes. It will be ready once CI passes

**Pull request merged**